compiler/rustc_parse/src/parser/item.rs

```diff
@@ -71,20 +73,50 @@ impl<'a> Parser<'a> {
         }
 
         if !self.eat(term) {
-            let token_str = super::token_descr(&self.token);
+            // The last token should be `term`: either EOF or `}`. If it's not that means that we've had an error
+            // parsing an item
             if !self.maybe_consume_incorrect_semicolon(items.last().map(|x| &**x)) {
-                let msg = format!("expected item, found {token_str}");
-                let mut err = self.dcx().struct_span_err(self.token.span, msg);
-                let span = self.token.span;
-                if self.is_kw_followed_by_ident(kw::Let) {
-                    err.span_label(
-                        span,
-                        "consider using `const` or `static` instead of `let` for global variables",
-                    );
-                } else {
-                    err.span_label(span, "expected item")
-                        .note("for a full list of items that can appear in modules, see <https://doc.rust-lang.org/reference/items.html>");
+                let token_str = super::token_descr(&self.token);
+                let mut err = self
+                    .dcx()
+                    .struct_span_err(self.token.span, format!("expected item, found {token_str}"));
+
+                match self.parse_full_stmt(AttemptLocalParseRecovery::No) {
+                    Ok(Some(stmt)) => {
+                        let span = stmt.span;
+                        match stmt.kind {
+                            StmtKind::Let(_) => {
+                                err.span_label(span, "unexpected `let` binding outside of a function")
+                                    .help("consider using `const` or `static` instead of `let` for global variables");
+                            }
+
+                            StmtKind::Semi(e) | StmtKind::Expr(e) => {
+                                err.span_label(span, "unexpected expression").help(format!(
+                                    "consider putting it inside a function: fn main() {{ {} }}",
+                                    pprust::expr_to_string(&*e)
+                                ));
+                            }
+                            StmtKind::MacCall(_) => {
+                                err.span_label(span, "unexpected macro call at this position");
```

Comment:
We should add examples for it
```
println!("hello");
```

Reply:
you mean `consider putting it inside a function`?